### PR TITLE
Implementation of per-channel fee setting. 

### PR DESCRIPTION
**Pull request by lullis**
Fixes: #4755 

## Description

This PR adds a "fee_schedule" attribute to `ChannelSchema` and exposes this for the GET/PUT/PATCH methods on all of the channel-related endpoints. This means that, by sending a PATCH request to the "api/v1/channels/<token_address>/<partner_address>" endpoint, the user can define the `flat` and `proportional` parts of the fee schedule.

NOTE: the `imbalance_fee` attribute is currently read-only and it is using the formula to calculate imbalance fees as defined by the `MediationFeeConfig` class.

### CODE REVIEW

File: raiden/api/rest.py [hunk 1]
```diff
@@ -1219,6 +1221,37 @@ def _set_channel_reveal_timeout(
         result = self.channel_schema.dump(updated_channel_state)
         return api_response(result=result)
 
+    def _set_fee_schedule(
+        self,
+        registry_address: TokenNetworkRegistryAddress,
+        channel_state: NettingChannelState,
+        fee_schedule: FeeScheduleState,
+    ):
+        log.debug("Set fee schedule")
```
Comment (rakanalh) [inline raiden/api/rest.py]: Remove this

File: raiden/api/rest.py [hunk 2]
```diff
@@ -1219,6 +1221,37 @@ def _set_channel_reveal_timeout(
...
+        if channel.get_status(channel_state) != ChannelState.STATE_OPENED:
```
Comment (rakanalh) [inline raiden/api/rest.py]: this validation should be done in `python.py`

Comment (rakanalh) [reply raiden/api/rest.py]: From there you can raise `UnexpectedChannelState`

File: raiden/api/rest.py [hunk 3]
```diff
@@ -1219,6 +1221,37 @@ def _set_channel_reveal_timeout(
...
+            return api_error(errors=str(e), status_code=HTTPStatus.BAD_REQUEST)
+        except InvalidFeeSchedule as e:
+            return api_error(errors=str(e), status_code=HTTPStatus.CONFLICT)
+
+        updated_channel_state = self.raiden_api.get_channel(
+            registry_address, channel_state.token_address, channel_state.partner_state.address
+        )
+
+        result = self.channel_schema.dump(updated_channel_state)
+        return api_response(result=result)
```
Comment (rakanalh) [inline raiden/api/rest.py]: You should not dump the schema manually. Return the channel state similar to other API handlers.

File: raiden/raiden_service.py [hunk 4]
```diff
@@ -998,7 +999,7 @@ def _initialize_channel_fees(self) -> None:
                 token_address=token_address,
             )
 
-            for channel in channels:
+            for channel in [c for c in channels if c.fee_schedule is None]:
```
Comment (rakanalh) [inline raiden/raiden_service.py]: Simplify here. Maybe add the comprehension to a variable before this line.

Comment (rakanalh) [reply raiden/raiden_service.py]: Also, document why we do this.

File: raiden/api/rest.py [hunk 5]
```diff
@@ -1219,6 +1221,37 @@ def _set_channel_reveal_timeout(
...
+        fee_schedule: FeeScheduleState,
+    ):
+        log.debug("Set fee schedule")
+        if channel.get_status(channel_state) != ChannelState.STATE_OPENED:
+            return api_error(
+                errors="Can't update the fee schedule of a closed channel",
+                status_code=HTTPStatus.CONFLICT,
+            )
+        try:
+            self.raiden_api.set_fee_schedule(
```
Comment (rakanalh) [inline raiden/api/rest.py]: I don't see any validation for the fee schedule values. Maybe @palango can advise on what rules that need to stand true before we dispatch the state change.

File: raiden/api/rest.py (hunk 3 above)
Comment (lullis) [reply raiden/api/rest.py]: I don't follow... This is following the exact same pattern as all of the private functions called inside `patch_channel`.

File: raiden/raiden_service.py (hunk 4 above)
Comment (lullis) [reply raiden/raiden_service.py]: All the function does is `_initialize_channel_fees`, but ok.

File: raiden/api/rest.py (hunk 5 above)
Comment (lullis) [reply raiden/api/rest.py]: The rules for these were sorted of punted, you are right. But for the flat/proportional attributes the validations are being done by marshmallow at the rest.py level. Perhaps they should also be added on the python.py side?

File: raiden/api/rest.py [hunk 6]
```diff
@@ -1270,74 +1300,52 @@ def patch_channel(
...
-                errors=(
-                    "Can not update a channel's reveal timeout "
-                    "and total deposit at the same time"
-                ),
-                status_code=HTTPStatus.CONFLICT,
-            )
+        attrs = locals()
+        changed_attributes = {
+            attr: attrs[attr] for attr in settable_attributes if attrs[attr] is not None
+        }
```
Comment (konradkonrad) [inline raiden/api/rest.py]: you should be able to achieve the same thing without `locals()` magic:
```python
patch_arguments = dict(
  total_deposit=total_deposit,
  total_withdraw=total_withdraw,
  reveal_timeout=reveal_timeout,
  fee_schedule=fee_schedule,
  state=state,
)
changed_attributes = {name: value for name, value in patch_arguments.items() if value is not None}
```

File: raiden/api/rest.py [hunk 7]
```diff
@@ -1270,74 +1300,52 @@ def patch_channel(
...
+                errors="Amount to withdraw must not be negative.",
+                status_code=HTTPStatus.BAD_REQUEST,
             )
 
-        empty_request = (
-            total_deposit is None
-            and state is None
-            and total_withdraw is None
-            and reveal_timeout is None
+        empty_request = all(
```
Comment (konradkonrad) [inline raiden/api/rest.py]: this is equivalent to `empty_request = len(changed_attributes) == 0`

File: raiden/api/rest.py [hunk 8]
```diff
@@ -1270,74 +1300,52 @@ def patch_channel(
...
         )
         if empty_request:
+            attribute_name_list = ", ".join((f"'{attr}'" for attr in settable_attributes))
             return api_error(
-                errors=(
-                    "Nothing to do. Should either provide "
-                    "'total_deposit', 'total_withdraw', 'reveal_timeout' or "
-                    "'state' argument"
-                ),
+                errors=("Nothing to do. Should provide one of {attribute_name_list} arguments"),
```
Comment (konradkonrad) [inline raiden/api/rest.py]: english:
```suggestion
                errors=("Nothing to do. Should provide one of the arguments: {attribute_name_list}."),
```

File: raiden/tests/integration/test_raidenservice.py [hunk 9]
```diff
@@ -135,8 +120,8 @@ def num_matching_queued_messages(room: str, message_type: Type) -> int:
             )
 
         assert num_matching_queued_messages(MONITORING_BROADCASTING_ROOM, RequestMonitoring) == 1
-        assert num_matching_queued_messages(PATH_FINDING_BROADCASTING_ROOM, PFSFeeUpdate) == 1
-        assert num_matching_queued_messages(PATH_FINDING_BROADCASTING_ROOM, PFSCapacityUpdate) == 1
+        assert num_matching_queued_messages(PATH_FINDING_BROADCASTING_ROOM, PFSFeeUpdate) == 0
```
Comment (konradkonrad) [inline raiden/tests/integration/test_raidenservice.py]: I don't understand, why this PR would change the messages exchanged? And if it does, then the comment from lines `103-107` must be updated!

File: raiden/api/v1/encoding.py [hunk 10]
```diff
@@ -303,6 +318,7 @@ class ChannelPutSchema(BaseSchema):
     reveal_timeout = fields.Integer(missing=None)
     settle_timeout = fields.Integer(missing=None)
     total_deposit = fields.Integer(default=None, missing=None)
+    fee_schedule = fields.Nested(FeeScheduleSchema)
```
Comment (konradkonrad) [inline raiden/api/v1/encoding.py]: does this make the fee schedule mandatory at channel creation?

File: raiden/api/rest.py [hunk 11]
```diff
@@ -1353,10 +1361,10 @@ def patch_channel(
                 errors="Requested channel for token {} and partner {} not found".format(
                     to_checksum_address(token_address), to_checksum_address(partner_address)
                 ),
-                status_code=HTTPStatus.CONFLICT,
+                status_code=HTTPStatus.NOT_FOUND,
```
Comment (konradkonrad) [inline raiden/api/rest.py]: a change of error codes must be reflected in the API documentation.
Actually this whole PR must be documented there: https://github.com/raiden-network/raiden/blob/develop/docs/rest_api.rst#channel-management

File: raiden/api/rest.py [hunk 12]
```diff
@@ -1270,74 +1300,52 @@ def patch_channel(
...
-                ),
+                errors=f"Can not update channel's {attribute_name_list} at the same time",
                 status_code=HTTPStatus.CONFLICT,
             )
 
         if total_deposit and total_deposit < 0:
             return api_error(
-                errors="Amount to deposit must not be negative.", status_code=HTTPStatus.CONFLICT
+                errors="Amount to deposit must not be negative.",
+                status_code=HTTPStatus.BAD_REQUEST,
```
Comment (konradkonrad) [inline raiden/api/rest.py]: see other comment about change of error codes

File: raiden/api/rest.py [hunk 13]
```diff
@@ -1270,74 +1300,52 @@ def patch_channel(
...
             )
 
         if total_withdraw and total_withdraw < 0:
             return api_error(
-                errors="Amount to withdraw must not be negative.", status_code=HTTPStatus.CONFLICT
+                errors="Amount to withdraw must not be negative.",
+                status_code=HTTPStatus.BAD_REQUEST,
```
Comment (konradkonrad) [inline raiden/api/rest.py]: see other comment about change of error codes

File: raiden/raiden_service.py [hunk 14]
```diff
@@ -418,7 +413,6 @@ def start(self) -> None:
         self._initialize_transactions_queues(chain_state)
         self._initialize_messages_queues(chain_state)
         self._initialize_whitelists(chain_state)
-        self._initialize_channel_fees()
```
Comment (konradkonrad) [inline raiden/raiden_service.py]: I see, you're removing setting channel fees at startup / restart. How do you persist the channel fee settings? How do you make sure, that channel fee settings are (or have been!) communicated to the selected PFS?

**Pull request closed without merging**